linear_operator/operators/block_matrix_linear_operator.py

```diff
@@ -0,0 +1,168 @@
+        self.num_tasks = len(self.linear_operators)
+        self.block_rows = linear_operators[0][0].shape[0]
+        self.block_cols = linear_operators[0][0].shape[1]
+
+        # Check that provided operators all have the same shape
+        T = self.num_tasks
+        for i in range(T):
+            for j in range(T):
+                assert (
+                    linear_operators[i][j].shape[0] == self.block_rows
+                ), "the number of rows much match for all linear operators"
+                assert (
+                    linear_operators[i][j].shape[1] == self.block_cols
+                ), "the number of columns much match for all linear operators"
+
+    @staticmethod
+    def create_square_ops_output(T: int) -> List[List[LinearOperator]]:
+        """Return an empty (square) list of operators of shape TxT"""
+        ops = []
+        for i in range(T):
+            tmp = []
+            for j in range(T):
+                tmp.append([])
+            ops.append(tmp)
+        return ops
+
+    def _matmul(
+        self: Float[LinearOperator, "*batch M N"],
+        rhs: Union[Float[torch.Tensor, "*batch2 N C"], Float[torch.Tensor, "*batch2 N"]],
+    ) -> Union[Float[torch.Tensor, "... M C"], Float[torch.Tensor, "... M"]]:
+
+        T = self.num_tasks
+
+        # A is block [N * T1, M * T2] and B is block [O * S1, P * S2]. If A and B have conformal block counts
+        # ie T2==S1 as well as M==O then use the blockwise algorithm. Else use to_dense()
+        if isinstance(rhs, self.__class__) and self.num_tasks == rhs.num_tasks and self.block_cols == rhs.block_rows:
+            output = BlockMatrixLinearOperator.create_square_ops_output(T)
+            for i in range(T):
+                for j in range(T):
+                    out_ij = self.linear_operators[i][0] @ rhs.linear_operators[0][j]
+                    for k in range(1, T):
+                        out_ij += self.linear_operators[i][k] @ rhs.linear_operators[k][j]
+                    output[i][j] = out_ij
+            return self.__class__(output)
+        elif isinstance(rhs, Tensor) and rhs.ndim == 2:
+            # Check both matrix dims divisible by T,
+            # reshape to (T, T, ), call block multiplication
+            if rhs.size(0) % T == 0 and rhs.size(1) % T == 0:
+                # A is block [N * T, M * T] and B is a general tensor/operator of shape [O, P].
+                # If O and P are both divisible by T,
+                # then interpret B as a [O//T * T, P//T * T] block matrix
+                O_T = rhs.size(0) // T
+                P_T = rhs.size(1) // T
+                rhs_blocks_raw = rhs.reshape(T, O_T, T, P_T)
+                rhs_blocks = rhs_blocks_raw.permute(0, 2, 1, 3)
+                rhs_op = BlockMatrixLinearOperator.from_tensor(rhs_blocks, T)
+                return self._matmul(rhs_op)
+
+        # Failover implementation. Convert to dense and multiply matricies
+        A = self.to_dense()
+        B = rhs.to_dense()
```

Comment:
`rhs` is a tensor, so it shouldn't have a `to_dense` method?

Reply:
Oh I see... rhs does get converted into a linear operator. (The type hit should probably be updated)
This will probably fail when `rhs` is a tensor but `rhs.dim() > 2`. It may be wise to do some type checking here.

Reply:
OK. I have updated the type signature. I haven't added explicit type-checking here since I expect the matrix multiplication of A@B to give an error message if the dimension sizes don't line up. (I think that explicit checking may just get unnecessarily messy here).